Add support for `EVENT_SHORT_REGION`

Krebrov001 opened the pull request:
- Relates to #1766

fkeceli committed:
Added the missing SHORT_REGION record type.

Signed-off-by: Fuat Keceli <fuat.keceli@intel.com>

Krebrov001 committed:
Add support for `EVENT_SHORT_REGION` to `ProfileTracer.cpp`

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

Krebrov001 committed:
Update the tests to reflect the `EVENT_SHORT_REGION` change

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

lhlawson commented on src/ProfileTracer.cpp:
```diff
@@ -78,6 +78,10 @@ namespace geopm
                     break;
+                case EVENT_SHORT_REGION:
+                    // integer value of number of short regions
+                    result = string_format_integer(value);
```
```suggestion
                    result = string_format_hex(value);
```
This should provide the hex value of region as it is written.  
We may want to add an EVENT_SHORT_REGION count (which would use the string_format_integer), but for #1766 only the region addition is needed.

Krebrov001 committed:
Fix the `ProfileTracerTest.format` failure

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

lhlawson commented:
We will want to run this against a series of workloads to see if we can hit the case called out in #1766 where an exception would occur.  We've confirmed that without this fix PENNANT still encounters the error:
`Error: <geopm> Invalid argument: unsupported event type: 3: at src/record.cpp:28`
when run using --geopm-profile-trace

If this fix does not resolve that error further work is needed.  Otherwise LGTM.

Krebrov001 committed:
Update the `ProfileTracer` class

- Use the `ApplicationSampler` to get the true value of the `EVENT_SHORT_REGION`
- Update constructors to enable use of the `ApplicationSampler`

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

Krebrov001 committed:
Update the `ProfileTracerTest`

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

cmcantalupo commented on src/ProfileTracerImp.hpp:
```diff
@@ -36,7 +37,13 @@ namespace geopm
+            static std::string event_format(double value);
     };
+
+    // defintion of the static data member
```
Please move to ProfileTracer.cpp otherwise it could potentially end up in multiple compilation units.

cmcantalupo commented on test/ProfileTracerTest.cpp:
```diff
@@ -44,6 +46,9 @@ void ProfileTracerTest::SetUp(void)
         time += 1.0;
     }
+
+     m_data.push_back({40, 0, geopm::EVENT_SHORT_REGION, 0xdeadbeef});
```
```suggestion
     m_data.push_back({40, 0, geopm::EVENT_SHORT_REGION, 88});
```
in general the event number will not match the region hash

cmcantalupo commented on test/ProfileTracerTest.cpp:
```diff
@@ -44,6 +46,9 @@ void ProfileTracerTest::SetUp(void)
     }
+
+     m_data.push_back({40, 0, geopm::EVENT_SHORT_REGION, 0xdeadbeef});
+     m_data.push_back({41, 1, geopm::EVENT_EPOCH_COUNT, region_hash});
```
```suggestion
     m_data.push_back({41, 1, geopm::EVENT_EPOCH_COUNT, 1});
```
The epoch count event should reflect a number times the epoch event has occurred.

cmcantalupo commented on test/ProfileTracerTest.cpp:
```diff
@@ -60,10 +65,17 @@ TEST_F(ProfileTracerTest, construct_update_destruct)
 
 TEST_F(ProfileTracerTest, format)
 {
+    EXPECT_CALL(m_application_sampler, get_short_region(0xdeadbeef))
```
```suggestion
    EXPECT_CALL(m_application_sampler, get_short_region(88))
```

cmcantalupo commented on test/ProfileTracerTest.cpp:
```diff
@@ -77,6 +89,8 @@ TEST_F(ProfileTracerTest, format)
         "36|1|REGION_EXIT|0xfa5920d6",
         "37|0|REGION_EXIT|0xfa5920d6",
+        "40|0|EVENT_SHORT_REGION|0xdeadbeef",
+        "41|1|EPOCH_COUNT|4200145110"
```
```suggestion
        "41|1|EPOCH_COUNT|1"
```

cmcantalupo reviewed (changes requested):
Just a couple of minor comments.

Looks like some issue with clang 12:
```
 FAIL: test/gtest_links/ProfileTracerTest.construct_update_destruct
 ```
 showed up in CI

cmcantalupo committed:
Fix up unit test parameter values to reflect real use case better

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Move static definition into cpp file

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>
